cql3/functions/functions.cc

```diff
@@ -53,6 +53,13 @@ bool abstract_function::requires_thread() const { return false; }
 
 bool as_json_function::requires_thread() const { return false; }
 
+static bool same_signature(shared_ptr<function> f1, shared_ptr<function> f2) {
```

Comment:
If you must use shared_ptr here - which I'm still hoping that you don't -, you can use use "const shared_ptr<function>& f1" here - this is a short-lived function that doesn't save the shared pointer, so it doesn't need to get an incremented reference count. Alternatively, you can just take "const function& f1" here, and ask the caller to dereference the shared pointer (which also doesn't need to touch the reference count).

Reply:
I don't necessarily have to use `shared_ptr` here, but it has an additional benefit over a `const function&` - we can test them for `nullptr` here and otherwise we would need to do that each time before calling this function.
I didn't think increasing the reference count is a big cost here, but the `const shared_ptr&` variant is strictly better so I'll use it.